src/compiler/checker.ts

```diff
@@ -36973,29 +36973,31 @@ namespace ts {
             }
 
             // Resolve the symbol as a value to ensure the type can be reached at runtime during emit.
-            const valueSymbol = resolveEntityName(typeName, SymbolFlags.Value, /*ignoreErrors*/ true, /*dontResolveAlias*/ false, location);
+            const valueSymbol = resolveEntityName(typeName, SymbolFlags.Value, /*ignoreErrors*/ true, /*dontResolveAlias*/ true, location);
+            const isTypeOnly = valueSymbol?.declarations?.every(isTypeOnlyImportOrExportDeclaration) || false;
+            const resolvedSymbol = valueSymbol && valueSymbol.flags & SymbolFlags.Alias ? resolveAlias(valueSymbol) : valueSymbol;
 
             // Resolve the symbol as a type so that we can provide a more useful hint for the type serializer.
             const typeSymbol = resolveEntityName(typeName, SymbolFlags.Type, /*ignoreErrors*/ true, /*dontResolveAlias*/ false, location);
-            if (valueSymbol && valueSymbol === typeSymbol) {
+            if (resolvedSymbol && resolvedSymbol === typeSymbol) {
                 const globalPromiseSymbol = getGlobalPromiseConstructorSymbol(/*reportErrors*/ false);
-                if (globalPromiseSymbol && valueSymbol === globalPromiseSymbol) {
+                if (globalPromiseSymbol && resolvedSymbol === globalPromiseSymbol) {
                     return TypeReferenceSerializationKind.Promise;
                 }
 
-                const constructorType = getTypeOfSymbol(valueSymbol);
+                const constructorType = getTypeOfSymbol(resolvedSymbol);
                 if (constructorType && isConstructorType(constructorType)) {
-                    return TypeReferenceSerializationKind.TypeWithConstructSignatureAndValue;
+                    return isTypeOnly ? TypeReferenceSerializationKind.TypeWithCallSignature : TypeReferenceSerializationKind.TypeWithConstructSignatureAndValue;
                 }
             }
 
             // We might not be able to resolve type symbol so use unknown type in that case (eg error case)
             if (!typeSymbol) {
-                return TypeReferenceSerializationKind.Unknown;
+                return isTypeOnly ? TypeReferenceSerializationKind.ObjectType : TypeReferenceSerializationKind.Unknown;

```

Comment:
How do we know the type is objecty? You can `import type` a primitive.

Reply:
`TypeSerializationKind.ObjectType` tells the `ts` transform to emit `Object` for the type metadata (which essentially means `any`). `TypeSerializationKind.Unknown` means we should try to resolve a value at runtime using that name (by using `typeof` testing), which would be incorrect behavior. Consider the following using the current version of the compiler (without this change):

```ts
// ts
import { dec } from "./decorators";
import type { Map } from "./myMap";

@dec
export class C {
  constructor(map: Map) {}
}

// js
...
let C = /** @class */ (() => {
    var _a;
    let C = class C {
        constructor(map) { }
    };
    C = __decorate([
        dec,
        __metadata("design:paramtypes", [typeof (_a = typeof Map !== "undefined" && Map) === "function" ? _a : Object])
    ], C);
    return C;
})();
```

The `design:paramtypes` metadata for `C` will *incorrectly* use the global `Map` object.

For a type-only import we can be *certain* that the type cannot be referenced as a value, so we should emit `Object` instead of trying to resolve a value for it.